File: pos_app/lib/features/products/data/products_repository.dart
// Offline-first product repository. Reads always hit the local Drift cache
// (sub-millisecond, works offline); refreshFromRemote() pulls the catalog when
// online. The full delta-sync engine (Part 6) supersedes the naive refresh.
import 'package:drift/drift.dart';

import '../../../core/network/connectivity_service.dart';
import '../../../data/local/database.dart';
import 'products_remote_ds.dart';

class ProductsRepository {
  final AppDatabase _db;
  final ProductsRemoteDataSource _remote;
  final ConnectivityService _connectivity;

  ProductsRepository(this._db, this._remote, this._connectivity);

  CatalogDao get _dao => _db.catalogDao;

  Future<Product?> byBarcode(String barcode) => _dao.byBarcode(barcode);
  Future<Product?> byId(String id) => _dao.byId(id);
  Future<List<Product>> search(String term) => _dao.search(term);
  Future<List<Product>> all() => _dao.allProducts();
  Stream<List<Product>> watch() => _dao.watchProducts();

  Future<int> localCount() async => (await _dao.allProducts()).length;

  /// Pull the catalog from the backend into the local cache. Returns the number
  /// of products written, or -1 if offline / the request failed.
  Future<int> refreshFromRemote() async {
    if (!_connectivity.isOnline) return -1;
    try {
      final rows = await _remote.fetchAll();
      final companions = rows.map(_toCompanion).toList();
      await _dao.upsertAllProducts(companions);
      return companions.length;
    } catch (_) {
      return -1;
    }
  }

  ProductsCompanion _toCompanion(Map<String, dynamic> j) {
    return ProductsCompanion(
      id: Value(j['id'] as String),
      sku: Value(j['sku'] as String?),
      barcode: Value(j['barcode'] as String?),
      name: Value(j['name'] as String),
      categoryId: Value(j['categoryId'] as String?),
      brandId: Value(j['brandId'] as String?),
      unitId: Value(j['unitId'] as String?),
      hsn: Value(j['hsn'] as String?),
      gstRate: Value((j['gstRate'] ?? 0) as int),
      purchasePrice: Value((j['purchasePrice'] ?? 0) as int),
      sellingPrice: Value((j['sellingPrice'] ?? 0) as int),
      mrp: Value((j['mrp'] ?? 0) as int),
      stock: Value((j['stock'] ?? 0) as int),
      reorderLevel: Value((j['reorderLevel'] ?? 0) as int),
      batchNo: Value(j['batchNo'] as String?),
      expiryAt: Value(j['expiryAt'] as int?),
      imageUrl: Value(j['imageUrl'] as String?),
      active: Value((j['active'] ?? true) as bool),
      updatedAt: Value((j['updatedAt'] ?? 0) as int),
    );
  }
}

File: backend/src/modules/products/repositories/products.repository.js
import { getDb } from '../../../db/index.js';

const COLS = `id, sku, barcode, name, category_id, brand_id, unit_id, hsn, gst_rate,
  purchase_price, selling_price, mrp, stock, reorder_level, batch_no, expiry_at,
  image_url, active, created_at, updated_at, deleted_at`;

export const ProductsRepository = {
  list({ limit = 100, offset = 0 } = {}) {
    return getDb()
      .prepare(`SELECT ${COLS} FROM products WHERE deleted_at IS NULL ORDER BY name LIMIT ? OFFSET ?`)
      .all(limit, offset);
  },
  findById(id) {
    return getDb().prepare(`SELECT ${COLS} FROM products WHERE id = ? AND deleted_at IS NULL`).get(id);
  },
  findByBarcode(barcode) {
    return getDb()
      .prepare(`SELECT ${COLS} FROM products WHERE barcode = ? AND deleted_at IS NULL`)
      .get(barcode);
  },
  search(term, limit = 25) {
    const like = `%${term}%`;
    return getDb()
      .prepare(
        `SELECT ${COLS} FROM products
         WHERE deleted_at IS NULL AND (name LIKE ? OR sku LIKE ? OR barcode = ?)
         ORDER BY name LIMIT ?`,
      )
      .all(like, like, term, limit);
  },
  insert(p) {
    getDb()
      .prepare(
        `INSERT INTO products (${COLS})
         VALUES (@id,@sku,@barcode,@name,@category_id,@brand_id,@unit_id,@hsn,@gst_rate,
                 @purchase_price,@selling_price,@mrp,@stock,@reorder_level,@batch_no,@expiry_at,
                 @image_url,@active,@created_at,@updated_at,@deleted_at)`,
      )
      .run(p);
    return this.findById(p.id);
  },
  update(id, p) {
    getDb()
      .prepare(
        `UPDATE products SET sku=@sku, barcode=@barcode, name=@name, category_id=@category_id,
           brand_id=@brand_id, unit_id=@unit_id, hsn=@hsn, gst_rate=@gst_rate,
           purchase_price=@purchase_price, selling_price=@selling_price, mrp=@mrp,
           stock=@stock, reorder_level=@reorder_level, batch_no=@batch_no, expiry_at=@expiry_at,
           image_url=@image_url, active=@active, updated_at=@updated_at
         WHERE id=@id AND deleted_at IS NULL`,
      )
      .run({ ...p, id });
    return this.findById(id);
  },
  softDelete(id, ts) {
    getDb().prepare('UPDATE products SET deleted_at = ?, updated_at = ? WHERE id = ?').run(ts, ts, id);
  },
  lowStock() {
    return getDb()
      .prepare(
        `SELECT ${COLS} FROM products
         WHERE deleted_at IS NULL AND stock <= reorder_level AND stock > 0 ORDER BY stock`,
      )
      .all();
  },
  outOfStock() {
    return getDb()
      .prepare(`SELECT ${COLS} FROM products WHERE deleted_at IS NULL AND stock <= 0 ORDER BY name`)
      .all();
  },
};
